tools/AutomationTestFramework/Management/RunbookProcessor.ps1

```diff
@@ -0,0 +1,207 @@
+                } $testListTemplate {
+                    Create-TestContent $folder.Path | Out-String
+                } default {
+                    $_
+                }
+            }
+            $content | Add-Content $bookPath
+        }
+        Write-Verbose "$bookPath generated."
+    }
+}
+
+# https://www.codeisahighway.com/how-to-easily-and-silently-obtain-accesstoken-bearer-from-an-existing-azure-powershell-session/
+function Get-AzureRmAccessToken() {
+    $azureRmProfile = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile
+    if(-not $azureRmProfile.Accounts.Count) {
+        Write-Error "Ensure you have logged in before calling this function."
+        return
+    }
+  
+    $currentAzureContext = Get-AzureRmContext
+    $profileClient = New-Object Microsoft.Azure.Commands.ResourceManager.Common.RMProfileClient($azureRmProfile)
+    Write-Debug ("Getting access token for tenant" + $currentAzureContext.Subscription.TenantId)
+    $token = $profileClient.AcquireAccessToken($currentAzureContext.Subscription.TenantId)
+    $token.AccessToken
+}
+
+function Start-Runbooks ([hashtable] $automation, [string] $runbooksPath) {
+    # Remove any previous jobs from this PS session
+    Get-Job | Remove-Job
+
+    foreach ($runbook in Get-ChildItem $runbooksPath) {
+        $bookName = $runbook.BaseName
+        Write-Verbose "Uploading '$bookName' runbook..."
+        $null = Import-AzureRmAutomationRunbook -Path $runbook.FullName -Name $bookName -type PowerShell -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -LogVerbose $true -Force -ErrorAction Stop
+        Write-Verbose "Publishing '$bookName' runbook..."
+        $null = Publish-AzureRmAutomationRunbook -Name $bookName -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -ErrorAction Stop
+        
+        $accessToken = Get-AzureRmAccessToken
+        $accountId = $(Get-AzureRmContext).Account.Id
+        $subscriptionName = $(Get-AzureRmContext).Subscription.Name
+        Write-Verbose "Starting '$bookName' runbook..."
+        $null = Start-Job `
+            -Name $bookName `
+            -ArgumentList $accessToken,$accountId,$subscriptionName,$bookName,$automation `
+            -ScriptBlock {
+            param($accessToken,$accountId,$subscriptionName,$bookName,$automation)
+            Login-AzureRmAccount -AccessToken $accessToken -AccountId $accountId -Subscription $subscriptionName
+            Start-AzureRmAutomationRunbook -Name $bookName -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -ErrorAction Stop -Wait -MaxWaitSeconds 3600
+        }
+        Write-Verbose "$bookName started."
+    }
+}
+
+function Wait-RunbookResults ([hashtable] $automation) {
+    Write-Verbose 'Waiting for runbooks to complete...'
+    $failedJobs = Get-Job | Wait-Job | ForEach-Object {
```

Comment:
It would be nice to actually have a list of jobs that you had created, this would keep you from waiting on other jobs that could be in the job repository in this session.

Reply:
I agree. But, after refactoring this entire automation process nearly 3 times, this was the solution that made sense at the time of originally implementing this. This wasn't my first solution either for getting results. Passing the job list out of the original automation process was not really feasible. 

With the current structure, I can change `Start-Runbooks` to return the Job IDs, and then have `Wait-RunbookResults` take in the list of Job IDs.

Reply:
You can wait on either job objects or job IDs, so whichever could be returned will work.